mmf/modules/attention.py

```diff
@@ -156,3 +160,107 @@ def forward(self, image_feat, question_embedding, image_locs=None):
                 masked_attention = self._mask_attentions(attention, image_locs)
 
         return masked_attention
+
+
+class MultiHeadAttention(nn.Module):
+    """
+    Multi-Head Attention implementation from https://arxiv.org/abs/1706.03762
+    """
+
+    def __init__(self, dim: int, num_attn: int, dropout: float = 0.1):
+        super().__init__()
+        self.p_attn = None
+        self.h = num_attn
+        self.d_k = dim // num_attn
+        self.linears = nn.ModuleList([nn.Linear(dim, dim) for _ in range(4)])
+        self.dropout = nn.Dropout(p=dropout)
+
+    def qkv_attention(
+        self,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
+        dropout: Type[nn.Dropout] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        d_k = query.size(-1)
+        scores = torch.matmul(query, key.transpose(-2, -1)) / math.sqrt(d_k)
+        if mask is not None:
+            scores.data.masked_fill_(mask.unsqueeze(1).unsqueeze(2), -1e9)
+
+        p_attn = nn.functional.softmax(scores, dim=-1)
+        if dropout is not None:
+            p_attn = dropout(p_attn)
+
+        return torch.matmul(p_attn, value), p_attn
+
+    def forward(
+        self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, mask: torch.Tensor
+    ) -> torch.Tensor:
+        b = q.size(0)
+
+        q = self.linears[0](q).view(b, -1, self.h, self.d_k).transpose(1, 2)
+        k = self.linears[1](k).view(b, -1, self.h, self.d_k).transpose(1, 2)
+        v = self.linears[2](v).view(b, -1, self.h, self.d_k).transpose(1, 2)
+
+        x, self.p_attn = self.qkv_attention(q, k, v, mask=mask, dropout=self.dropout)
+        x = x.transpose(1, 2).contiguous().view(b, -1, self.h * self.d_k)
+
+        return self.linears[-1](x)
+
+
+class SelfAttention(nn.Module):
```

Comment:
Same here, these are implemented for Transformer module in torch. Maybe have a look?